### PR TITLE
URL Regular Expression Modification

### PR DESCRIPTION
**Pull request by malvidin**
Fix catastrophic backtracking on URL (issue was for `re` only). The new expression uses a greedy range for all upper Unicode characters, but was duplicated somewhat because of the way `re2` accepts those ranges is different than `re` and `regex`. 

Add related tests for URL matches to address backtracking, unicode, and IPv6.

### CODE REVIEW

File: eml_parser/regexes.py [hunk 1]
```diff
@@ -37,7 +37,39 @@
 #   - do not use a fixed list of TLDs but rather \w
 #   - only check for URLs with scheme
 #   - modify the end marker to allow any acceptable char according to the RFC3986
-url_regex_simple = re.compile(r'''(?i)\b(?:https?|ftps?):(?:/{1,3}|[a-z0-9%])(?:[^\s()<>{}\[\]]+|\([^\s()]*?\([^\s()]+\)[^\s()]*?\)|\([^\s]+?\))+(?:[\w\-._~%!$&'()*+,;=:/?#\[\]@]+)''')
+
+try:
```
Comment (sim0nx) [inline eml_parser/regexes.py:41]: Could you please add an explanation as to why you include two slightly different re.compile blocks. Which one works/doesn't for what lib.

**Comment by sim0nx** [comment]
Thanks for the PR!

Have you thoroughly tested your regex to work as expected ? 
I am a bit reluctant with replacing that one.

**Comment by malvidin** [comment]
I understand the concern. I added a few tests to ensure that some URLs are matched, but the current tests are definitely not comprehensive.  

I would appreciate additional URL samples that shouldn't match the initial simple regex. The follow-on filtering by urllib should help decrease reliance that the regex is perfect.

**Comment by malvidin** [comment]
I tried to add "www.*" back into the expression, but it fails for re2. Without using a forward lookahead, "(?=X)", I do not know a good way to exclude email addresses in the expression.  Is there any desire to add this back to the expression?

```
(?:
    www\d{0,4}[.]
 (?:[-\w\u0900-\u2017\u2020-\U0010FFFF]{1,300}[.]){1,300}
    [-0-9a-z\u0900-\u0DFF]{2,30}[.]*
 (?=(?:[/#%?\s'"\u2018-\u201F]|$))
)
```

If "www.*" addresses were added, I don't see any additional required changes to the current urllib validation. For re2, it would probably require modification to url_regex_simple to exclude email addresses, get_uri_ondata to include email addresses that start with www.

File: eml_parser/regexes.py (hunk 1 above)
Comment (malvidin) [reply eml_parser/regexes.py]: I have changed the block from a try/except to if/else based on the re.__name__. `re2` uses `\x{0000}` instead of `\u0000` and '\U00000000`.  This is a simple enough change that one string could be compiled after a re.sub() modifies the string, but there are other implementation differences between re, re2, and regex that may require entirely separate expressions.

**Comment by sim0nx** [comment]
> I understand the concern. I added a few tests to ensure that some URLs are matched, but the current tests are definitely not comprehensive.
> 
> I would appreciate additional URL samples that shouldn't match the initial simple regex.

I appreciate the tests you included and the rather large list of URLs to test against :-)
I can't easily come up with more URLs to test with at this time unfortunately.

> The follow-on filtering by urllib should help decrease reliance that the regex is perfect.

Fully agreed

File: eml_parser/regexes.py [hunk 2]
```diff
@@ -37,7 +37,31 @@
 #   - do not use a fixed list of TLDs but rather \w
 #   - only check for URLs with scheme
 #   - modify the end marker to allow any acceptable char according to the RFC3986
-url_regex_simple = re.compile(r'''(?i)\b(?:https?|ftps?):(?:/{1,3}|[a-z0-9%])(?:[^\s()<>{}\[\]]+|\([^\s()]*?\([^\s()]+\)[^\s()]*?\)|\([^\s]+?\))+(?:[\w\-._~%!$&'()*+,;=:/?#\[\]@]+)''')
+
+if re.__name__ == 're2':
+    url_regex_simple = re.compile(r'''
```
Comment (sim0nx) [inline eml_parser/regexes.py:42]: Quickly testing some stuff I noticed that the following URL does not match with your modified regex:
http://example.com/?

Could you please have a look at that ?

-> FYI I tested with the non re2 version

**Comment by sim0nx** [comment]
> I tried to add "www.*" back into the expression, but it fails for re2. Without using a forward lookahead, "(?=X)", I do not know a good way to exclude email addresses in the expression. Is there any desire to add this back to the expression?
> 
> ```
> (?:
>     www\d{0,4}[.]
>  (?:[-\w\u0900-\u2017\u2020-\U0010FFFF]{1,300}[.]){1,300}
>     [-0-9a-z\w\u0900-\u0DFF]{2,30}[.]*
>  (?=(?:[/#%?\s'"\u2018-\u201F]|$))
> )
> ```
> 
> If "www.*" addresses were added, I don't see any additional required changes to the current urllib validation. For re2, it would probably require modification to url_regex_simple to exclude email addresses, get_uri_ondata to include email addresses that start with www.

Hmm I'm not sure what you mean by "adding back", there was no check for www.* or do I miss something ?

**Comment by malvidin** [comment]
By adding back, I was referencing the gist from gruber from the comments above the regex.

File: eml_parser/regexes.py (hunk 2 above)
Comment (malvidin) [reply eml_parser/regexes.py:42]: The regex should match `http://example.com/?`, but urllib.parse removes the trailing `?`

Comment (sim0nx) [reply eml_parser/regexes.py:42]: At least on regex101 it does not work, see https://regex101.com/r/om1AXp/1

Strangely "https://www.example.com?" matches

Comment (malvidin) [reply eml_parser/regexes.py:42]: I had `?` in the negated character class. I had removed the `?` for #69, but had not merged that into this.

Comment (sim0nx) [reply eml_parser/regexes.py:42]: looks good

**Comment by sim0nx** [comment]
> By adding back, I was referencing the gist from gruber from the comments above the regex.

Ok, no there is no desire for adding that back

**Comment by sim0nx** [comment]
Thanks for the PR!
